src/pub_macros.rs

```diff
@@ -0,0 +1,120 @@
+///
+/// // Information to translator abstract String type keys to concrete bitcoin::PublicKey.
+/// // In practice, wallets would map from String key names to BIP32 keys
+/// struct StrPkTranslator {
+///     pk_map: HashMap<String, bitcoin::PublicKey>
+/// }
+///
+/// // If we also wanted to provide mapping of other associated types(sha256, older etc),
+/// // we would use the general Translator Trait.
+/// impl Translator<String, bitcoin::PublicKey, ()> for StrPkTranslator {
+///     // Provides the translation public keys P -> Q
+///     fn pk(&mut self, pk: &String) -> Result<bitcoin::PublicKey, ()> {
+///         self.pk_map.get(pk).copied().ok_or(()) // Dummy Err
+///     }
+///
+///     // If our policy also contained other fragments, we could provide the translation here.
+///     fn pkh(&mut self, pkh: &String) -> Result<hash160::Hash, ()> {
+///         unreachable!("Policy does not contain any pkh fragment");
+///     }
+///
+///     // Fail for hash types
+///     translate_assoc_fail!(String, bitcoin::PublicKey, ());
+/// }
+///
+/// let mut pk_map = HashMap::new();
+/// pk_map.insert(String::from("alice_key"), bitcoin::PublicKey::from_str(alice_key).unwrap());
+/// pk_map.insert(String::from("bob_key"), bitcoin::PublicKey::from_str(bob_key).unwrap());
+/// let mut t = StrPkTranslator { pk_map: pk_map };
+/// ```
+#[macro_export]
+macro_rules! translate_assoc_fail {
+    ($source: ty, $target:ty, $error_ty: ty) => {
+        fn sha256(
+            &mut self,
+            _sha256: &<$source as $crate::MiniscriptKey>::Sha256,
+        ) -> Result<<$target as $crate::MiniscriptKey>::Sha256, $error_ty> {
+            panic!("Called sha256 on translate_only_pk")
+        }
+
+        fn hash256(
+            &mut self,
+            _hash256: &<$source as $crate::MiniscriptKey>::Hash256,
+        ) -> Result<<$target as $crate::MiniscriptKey>::Hash256, $error_ty> {
+            panic!("Called hash256 on translate_only_pk")
+        }
+
+        fn hash160(
+            &mut self,
+            _hash160: &<$source as $crate::MiniscriptKey>::Hash160,
+        ) -> Result<<$target as $crate::MiniscriptKey>::Hash160, $error_ty> {
+            panic!("Called hash160 on translate_only_pk")
+        }
+
+        fn ripemd160(
+            &mut self,
+            _ripemd160: &<$source as $crate::MiniscriptKey>::Ripemd160,
+        ) -> Result<<$target as $crate::MiniscriptKey>::Ripemd160, $error_ty> {
+            panic!("Called ripemd160 on translate_only_pk")
+        }
+    };
+}
+
+/// Macro for translation of associated types where the associated type is the same
+/// Handy for Derived -> concrete keys where the associated types are the same.
+///
+/// Writing the complete translator trait is tedious. This macro is handy when
+/// we are not trying the associated types for hash160, ripemd160, hash256 and
+/// sha256.
+///
+/// See also [`crate::translate_assoc_fail`]
+#[macro_export]
+macro_rules! translate_assoc_clone {
```

Comment:
What if we instead made this `translate_assoc_into` and called `.into()` rather than `.clone()`? This would work for all the hash types since they are `Copy`, and would potentially work more generally.

Also I think both these macros should somehow be renamed to indcate that they only implement the translate functions for hashes, not for pubkeys.

Reply:
Agreed about the renaming. I was also thinking of another refactor that removes translate_pkh altogether. Now that we have Pkh(Pk) instead of Pkh(Pk::Hash), I don't think it is needed. We should support any translation for abstract `RawPkh` until we have raw descriptors specified.

Reply:
Yeah, agreed, let's drop `translate_pkh`.